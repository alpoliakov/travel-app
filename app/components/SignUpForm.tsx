import Link from 'next/link';
import React, { useEffect, useState } from 'react';
import { useIntl } from 'react-intl';

import { REG_EMAIL } from '../constants';
import { useAuth } from '../lib/useAuth';
import Loader from './Loader';

export default function SignUpForm() {
  const [userData, setUserData] = useState({
    name: '',
    email: '',
    password: '',
    avatar: '',
  });
  const [loading, setLoading] = useState(false);
  const [errorPsw, setErrorPsw] = useState(false);
  const [errorName, setErrorName] = useState(false);
  const [validEmail, setValidEmail] = useState(false);
  const { user, signUp, error } = useAuth();

  const { formatMessage: f } = useIntl();

  const validation = async ({ name, email, password, avatar }) => {
    let result = false;
    if (!password || password.length < 6) {
      await setErrorPsw(true);
      result = true;
    }

    if (!name || name.length < 3) {
      await setErrorName(true);
      result = true;
    }

    if (!email || !email.match(REG_EMAIL)) {
      await setValidEmail(true);
      result = true;
    } else {
      await setValidEmail(false);
    }

    return result;
  };

  const onFinish = async (e) => {
    e.preventDefault();
    setLoading(true);
    const { name, email, password } = userData;
    let { avatar } = userData;

    const valid = await validation(userData);

    if (!avatar) {
      avatar = '/images/hacker.png';
    }

    if (valid) {
      setLoading(false);
      return;
    }

    await signUp(name, email, password, avatar);
  };

  useEffect(() => {
    if (user) {
      setLoading(false);
    }

    if (error) {
      setLoading(false);
    }
  }, [user, error]);

  return (
    <div className="w-full max-w-lg">
      <div className="leading-loose">
        <Loader show={loading} />
        <form className="m-4 p-10 bg-white bg-opacity-25 rounded shadow-xl">
          <p className="text-white font-medium text-center text-lg font-bold uppercase">
            {f({ id: 'register' })}
          </p>
          <div className="">
            <label className="block text-sm text-white" htmlFor="name">
              {f({ id: 'name' })}
            </label>
            <input
              className="w-full px-5 py-1 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
              type="name"
              id="name"
              value={userData.name}
              placeholder={f({ id: 'phName' })}
              aria-label="name"
              onChange={(e) => {
                setUserData({ ...userData, name: e.target.value });

                if (userData.name.length > 2) {
                  setErrorName(false);
                }
              }}
              required
            />
            <span className="flex items-center font-medium tracking-wide text-yellow-500 text-xs mt-1 ml-1">
              {errorName && f({ id: 'errorName' })}
            </span>
          </div>
          <div className="mt-1">
            <label className="block text-sm text-white" htmlFor="email">
              E-mail:
            </label>
            <input
              className="w-full px-5 py-1 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
              type="email"
              id="email"
              value={userData.email}
              placeholder={f({ id: 'phEmail' })}
              aria-label="email"
              onChange={(e) => setUserData({ ...userData, email: e.target.value })}
              required
            />
            <span className="flex items-center font-medium tracking-wide text-yellow-500 text-xs mt-1 ml-1">
              {validEmail && f({ id: 'errorEmail' })}
            </span>
          </div>
          <div className="mt-1">
            <label className="block  text-sm text-white" htmlFor="password">
              {f({ id: 'pass' })}
            </label>
            <input
              className="w-full px-5 py-1 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
              type="password"
              id="password"
              value={userData.password}
              onChange={(e) => {
                setUserData({ ...userData, password: e.target.value });
                if (userData.password.length > 5) {
                  setErrorPsw(false);
                }
              }}
              name="password"
              minLength={6}
              placeholder={f({ id: 'phPass' })}
              arial-label="password"
              required
            />
            <span className="flex items-center font-medium tracking-wide text-yellow-500 text-xs mt-1 ml-1">
              {errorPsw && f({ id: 'errorPass' })}
            </span>
          </div>
          <div className="mt-1">
            <label className="block  text-sm text-white" htmlFor="avatar">
              {f({ id: 'avatar' })}
            </label>
            <input
              className="w-full px-5 py-1 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
              type="avatar"
              id="avatar"
              value={userData.avatar}
              onChange={(e) => setUserData({ ...userData, avatar: e.target.value })}
              name="avatar"
              placeholder={f({ id: 'phAvatar' })}
              arial-label="loader"
            />
            <span className="flex items-center font-medium tracking-wide text-yellow-500 text-xs mt-1 ml-1">
              Invalid username field !
            </span>
          </div>

          <div className="mt-8 items-center flex justify-between">
            <button
              className="px-4 py-1 text-white font-bold tracking-wider bg-gray-900 hover:bg-gray-800 rounded"
              type="submit"
              onClick={onFinish}>
              {f({ id: 'enter' })}
            </button>
            <Link href="/auth/login">
              <a
                className="inline-block right-0 align-baseline font-bold text-sm text-500 text-white hover:text-blue-800"
                role="button">
                Or {f({ id: 'btnLogin' })}
              </a>
            </Link>
          </div>
        </form>
      </div>
    </div>
  );
}
