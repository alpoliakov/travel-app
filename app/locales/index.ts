export const messages = {
  en: {
    btnLogin: 'Login',
    btnSignUp: 'Sign Up',
    search: 'Search',
    phEmail: 'Enter your e-mail',
    phPass: 'Enter your password',
    createAcc: 'Create an account',
    return: 'Return ?',
    enter: 'Enter',
    load: 'Loading',
    high: 'Max',
    low: 'Min',
    now: 'Currently',
    pass: 'Password (6 characters minimum):',
    phName: 'Enter your name',
    phAvatar: 'Upload your avatar',
    name: 'Name (3 characters minimum):',
    avatar: 'Avatar:',
    register: 'Sign Up',
    signOut: 'Sign Out',
    errorEmail: 'Invalid email field!',
    errorPass: 'Invalid password field!',
    errorName: 'Invalid username field!',
    profile: 'Profile',
    delete: 'Delete',
    or: 'Or',
    selectFile: 'Select a file',
    capital: 'Capital',
    hello: 'Hello!',
    modalText: 'To view, you need to log in or register!',
    close: 'Close',
    ratingQuestion: 'How do you rate this place?',
    opinion: 'Your opinion matters to us!',
    rateButton: 'Rate now',
    later: 'Maybe later',
    reminder: 'You need to login or register!',
    feedbackGive: 'give feedback',
    feedbackSee: 'see reviews',
    usd: 'USD',
    eur: 'EUR',
    rub: 'RUB',
  },
  ru: {
    btnLogin: 'Войти',
    btnSignUp: 'Регистрация',
    search: 'Поиск',
    phEmail: 'Введите ваш e-mail',
    phPass: 'Введите ваш пароль',
    createAcc: 'Создать аккаунт',
    return: 'Вернуться ?',
    enter: 'Ввести',
    load: 'Загрузка',
    high: 'Макс',
    low: 'Мин',
    now: 'Сейчас',
    pass: 'Пароль (6 символов минимум):',
    phName: 'Введите ваше имя',
    phAvatar: 'Загрузите ваш аватар',
    name: 'Имя (3 символа минимум):',
    avatar: 'Аватар:',
    register: 'Регистрация',
    signOut: 'Выйти',
    errorEmail: 'Невалидное поле электронной почты!',
    errorPass: 'Невалидное поле пароля!',
    errorName: 'Невалидное поле имени пользователя!',
    profile: 'Профиль',
    delete: 'Удалить',
    or: 'Или',
    selectFile: 'Выберите файл',
    capital: 'Столица',
    hello: 'Приветствуем!',
    modalText: 'Для просмотра необходимо авторизоваться или зарегистрироваться!',
    close: 'Закрыть',
    ratingQuestion: 'Как вы оцениваете это место?',
    opinion: 'Ваше мнение важно для нас!',
    rateButton: 'Оценить сейчас',
    later: 'Может быть, позже',
    reminder: 'Вам необходимо войти или зарегистрироваться!',
    feedbackGive: 'оставить отзыв',
    feedbackSee: 'посмотреть отзывы',
    usd: 'Доллар',
    eur: 'Евро',
    rub: 'Рубль',
  },
  uk: {
    btnLogin: 'Увійти',
    btnSignUp: 'Реєстрація',
    search: 'Пошук',
    phEmail: 'Введіть ваш e-mail',
    phPass: 'Введіть ваш пароль',
    createAcc: 'Створити акаунт',
    return: 'Повернутись ?',
    enter: 'Ввести',
    load: 'Завантаження',
    high: 'Макс',
    low: 'Miн',
    now: 'Зараз',
    pass: 'Пароль (6 символів мінімум):',
    phName: 'Введіть ваше им`я',
    phAvatar: 'Завантажте ваш аватар',
    name: 'Ім`я (3 символа мінімум):',
    avatar: 'Аватар:',
    register: 'Реєстрація',
    signOut: 'Вийти',
    errorEmail: 'Невалідне поле електронної пошти!',
    errorPass: 'Невалідне поле пароля!',
    errorName: 'Невалідне поле імені користувача',
    profile: 'Профіль',
    delete: 'Видалити',
    or: 'Або',
    selectFile: 'Виберіть файл',
    capital: 'Столиця',
    hello: 'Вітаємо!',
    modalText: 'Для перегляду потрібно увійти або зареєструватися!',
    close: 'Закрити',
    ratingQuestion: 'Як ви оцінюєте це місце?',
    opinion: 'Ваша думка для нас важлива!',
    rateButton: 'Оцініть зараз',
    later: 'Можливо пізніше',
    reminder: 'Вам необхідно увійти або зареєструватися!',
    feedbackGive: 'залишити відгук',
    feedbackSee: 'подивитися відгуки',
    usd: 'Долар',
    eur: 'Євро',
    rub: 'Рубль',
  },
};
