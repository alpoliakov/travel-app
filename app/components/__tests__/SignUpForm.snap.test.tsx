import React from 'react';
import { IntlProvider } from 'react-intl';
import renderer from 'react-test-renderer';

import { messages } from '../../locales';
import SignUpForm from '../SignUpForm';

jest.mock('../../lib/useAuth', () => ({
  useAuth: jest.fn(() => ({ signIn: jest.fn(), user: '', error: '' })),
  AuthProvider: ({ children }) => <div>{children}</div>,
}));

it('SignUpForm renders correctly', () => {
  const tree = renderer
    .create(
      <IntlProvider locale="ru" messages={messages['ru']}>
        <SignUpForm />)
      </IntlProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
