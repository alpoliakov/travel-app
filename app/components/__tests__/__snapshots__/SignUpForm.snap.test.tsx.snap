// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm renders correctly 1`] = `
Array [
  <div
    className="w-full max-w-lg"
  >
    <div
      className="leading-loose"
    >
      <form
        className="m-4 p-10 bg-white bg-opacity-25 rounded shadow-xl"
      >
        <p
          className="text-white font-medium text-center text-lg font-bold uppercase"
        >
          Регистрация
        </p>
        <div
          className=""
        >
          <label
            className="block text-sm text-white"
            htmlFor="name"
          >
            Имя (3 символа минимум):
          </label>
          <input
            aria-label="name"
            className="w-full px-5 py-1 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
            id="name"
            onChange={[Function]}
            placeholder="Введите ваше имя"
            required={true}
            type="name"
            value=""
          />
          <span
            className="flex items-center font-medium tracking-wide text-yellow-500 text-xs mt-1 ml-1"
          />
        </div>
        <div
          className="mt-1"
        >
          <label
            className="block text-sm text-white"
            htmlFor="email"
          >
            E-mail:
          </label>
          <input
            aria-label="email"
            className="w-full px-5 py-1 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
            id="email"
            onChange={[Function]}
            placeholder="Введите ваш e-mail"
            required={true}
            type="email"
            value=""
          />
          <span
            className="flex items-center font-medium tracking-wide text-yellow-500 text-xs mt-1 ml-1"
          />
        </div>
        <div
          className="mt-1"
        >
          <label
            className="block  text-sm text-white"
            htmlFor="password"
          >
            Пароль (6 символов минимум):
          </label>
          <input
            arial-label="password"
            className="w-full px-5 py-1 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
            id="password"
            minLength={6}
            name="password"
            onChange={[Function]}
            placeholder="Введите ваш пароль"
            required={true}
            type="password"
            value=""
          />
          <span
            className="flex items-center font-medium tracking-wide text-yellow-500 text-xs mt-1 ml-1"
          />
        </div>
        <div
          className="mt-4 flex justify-between align-center"
        >
          <label
            className="block text-sm text-white"
            htmlFor="avatar"
          >
            Аватар:
            <div
              className="w-full cursor-pointer tracking-wide px-5 py-2 text-gray-700 bg-gray-300 rounded focus:outline-none focus:bg-white"
            >
              <input
                accept="image/*"
                arial-label="loader"
                className="hidden"
                id="avatar"
                name="file"
                onChange={[Function]}
                placeholder="Загрузите ваш аватар"
                type="file"
              />
              <span
                className="mt-2 text-base leading-normal"
              >
                Выберите файл
              </span>
            </div>
          </label>
          <div
            className="mr-3"
          >
            <img
              alt="avatar"
              className="h-20 w-20 rounded-full"
              src="/images/hacker.png"
            />
          </div>
        </div>
        <div
          className="mt-8 items-center flex justify-between"
        >
          <button
            className="px-4 py-1 text-white font-bold tracking-wider bg-gray-900 hover:bg-gray-800 rounded"
            onClick={[Function]}
            type="submit"
          >
            Ввести
          </button>
          <a
            className="inline-block right-0 align-baseline font-bold text-sm text-500 text-white hover:text-blue-800"
            href="/auth/login"
            onClick={[Function]}
            onMouseEnter={[Function]}
            role="button"
          >
            Или
             
            Войти
          </a>
        </div>
      </form>
    </div>
  </div>,
  ")",
]
`;
